ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt

```diff
@@ -12,35 +20,33 @@ internal class KopyKatProcessor(
 ) : SymbolProcessor {
 
   override fun process(resolver: Resolver): List<KSAnnotated> {
-    with(resolver.getAllFiles()) {
-      if (none { file -> file.hasGeneratedMarker() }) {
-        flatMap { file -> file.declarations }
+    resolver.getAllFiles().let { files ->
+      if (files.none { it.hasGeneratedMarker() }) {
+        val targets = files.flatMap { it.declarations }
           .filterIsInstance<KSClassDeclaration>()
-          .forEach {
-            logger.logging("Processing ${it.simpleName}", it)
-            it.process()
-          }
+          .filter { it.requiresProcessing() }
+        targets
+          .onEach { logger.logging("Processing ${it.simpleName}", it) }
+          .forEach { it.onClassScope(targets) { process() } }
       }
     }
     return emptyList()
   }
 
-
-  private fun KSClassDeclaration.process() {
+  private fun ClassScope.process() {
     when {
       options.hierarchyCopy && isSealedDataHierarchy() -> {
-        HierarchyCopyFunctionKt.writeTo(codegen)
-        if (options.copyMap) CopyMapFunctionKt.writeTo(codegen)
-        if (options.mutableCopy) MutableCopyKt.writeTo(codegen)
-      }
-      isDataClass() -> {
-        if (options.copyMap) CopyMapFunctionKt.writeTo(codegen)
-        if (options.mutableCopy) MutableCopyKt.writeTo(codegen)
+        if (options.copyMap) HierarchyCopyMapFunctionKt.writeTo(codegen)
+        if (options.mutableCopy) hierarchyCopyFunctionKt().writeTo(codegen)
       }
-      isValueClass() -> {
-        if (options.valueCopy) ValueCopyFunctionKt.writeTo(codegen)
```

Comment:
Where has this gone? I guess it's now part of `copyMap`?

Reply:
Yeah, I managed to merge both Value and Data classes into the same strategy. I tried to make the MutableCopy for value classes also be a value class, but value classes don't allow mutability, duh! :sweat\_smile: